python/src/iceberg/io/base.py

```diff
@@ -26,6 +26,50 @@
 from abc import ABC, abstractmethod
 from typing import Union
 
+from typing_extensions import Protocol, runtime_checkable
```

Comment:
might pay to put a conditional import here?

Reply:
i.e. try to import from the native 3.8+ version and fallback to extension?  I think this could make the typing_extensions optional for 3.8 (not sure if we consider this a big deal).

Reply:
That's a great catch, I updated this to:
```py
try:
    from typing import Protocol, runtime_checkable
except ImportError:  # pragma: no cover
    from typing_extensions import Protocol  # type: ignore
    from typing_extensions import runtime_checkable
```